### ISSUE TITLE
Course Continue Button directs to next incomplete lesson regardless of lesson availability

### ISSUE THREAD

**Issue by wonboyn** [gocodebox/lifterlms#728]
_EDIT: See [comment below](https://github.com/gocodebox/lifterlms/issues/728#issuecomment-451778479) for refined reproduction steps and issue clarification._

### 1. Reproduction Steps

+ Course configured with modules/sections each having a number of lessons
+ All lessons in a module/section released 7 days after the previous module (ie release one module of lessons a week)
+ Student has **_completed_** all lessons in module 1 and module 2
+ Next module not due to be released for 3 days
+ Student logs out & then logs in again
+ Student now unable to access any of the completed lessons & gets an error message on course landing page:
"The lesson "Creating the best conditions for sleep" will be available on December 30, 2018 4:47 pm"

<img width="1504" alt="screen shot 2018-12-27 at 1 07 20 pm" src="https://user-images.githubusercontent.com/8089290/50462291-5f6bd880-09d8-11e9-9644-09c732583026.png">

### 2. Expected Behavior

+ Student should be able to log in & access ANY of the lessons in modules that have been released
+ If a student tries to access a lesson in a module that has not been released THEN they get error message shown in the screenshot above - this error should be shown on the lesson view, NOT on the course landing page

### 3. Actual Behavior

+ Student is unable progress beyond the course landing page
+ Student is unable to access any lessons - regardless of whether they have been released or not

### 4. Error Messages / Logs

+ "The lesson "Creating the best conditions for sleep" will be available on December 30, 2018 4:47 pm"

### 6. Browser, Device, and Operating System

+ Safari Version 12.0.2 (14606.3.4) on OSX 

<img width="698" alt="screen shot 2018-12-27 at 1 10 48 pm" src="https://user-images.githubusercontent.com/8089290/50462346-da34f380-09d8-11e9-8996-0e08a9c8031c.png">

**Comment by thomasplevy** [comment]
@wonboyn would you please attach a screenshot of the settings for the lesson that's throwing this error message please

**Comment by thomasplevy** [comment]
@wonboyn 

Can you open a support request about this?

I feel like I need to dig into your enrollment data to get enough information to accurately recreate this. 

GitHub is not ideal for support.

Thanks

**Comment by wonboyn** [comment]
Hey Thomas,

Happy to - however when I log in to the Lifter site & click on the support tab I don’t have access to create a new ticket.
I only have a big orange button saying “Buy LifterLMS Pro Now”.

Tim

**Comment by thomasplevy** [comment]
@wonboyn 

Sounds like you don't have access to support then.

Can you provide me with an export file of this course so I can recreate the issue exactly.

Thanks

**Comment by wonboyn** [comment]
Here you go.

**Comment by thomasplevy** [comment]
@wonboyn try again please! nothing was attached / provided here

**Comment by wonboyn** [comment]
@thomasplevy 

Strange. Anyway here it is again.

[A-Mindful-Way-To-Healthy-Sleep-8211-Self-Supported_20181228.json.zip](https://github.com/gocodebox/lifterlms/files/2713551/A-Mindful-Way-To-Healthy-Sleep-8211-Self-Supported_20181228.json.zip)

One thing I forgot to mention.
We were alerted to this issue by a student who was blocked from accessing lessons he had already marked as complete. To allow him to continue we added one new, empty lesson to the end of section 2. This allowed him to log in & access the previous lessons.

This export includes that extra lesson in section 2

**Comment by thomasplevy** [comment]
@wonboyn,

A few notes after reviewing your course and setting up a reproduction:

> 2. Expected Behavior

> + Student should be able to log in & access ANY of the lessons in modules that have been released

It seems to me that it is possible to access completed lessons. Based on your screenshots it looks like you have a full width view with no sidebar our course outline. If the syllabus has been removed below the fold it may *feel* as if you can't access content but I bet if you have the student navigate to a direct URL for one of the completed lessons they will show up there without issue.

I'm guessing here but I think this is more of a UX issue with how you've setup your course layout with BB and the full width view that could be causing this issue.

> + If a student tries to access a lesson in a module that has not been released THEN they get error message shown in the screenshot above - this error should be shown on the lesson view, NOT on the course landing page

Lessons are 100% inaccessible in LifterLMS when the student doesn't have access. When a student attempts to access a URL (via a direct link) they are redirected to the course where that error message is displayed.

We can argue about whether or not this is proper behavior but it's not unexpected based on the way LifterLMS works. EG: this is not a bug. I wrote the code and I wrote it to intentionally redirect to the course and display the error message there.


Can you let me know if this helps explain the behavior you're encountering on your site?

**Comment by wonboyn** [comment]
Hi @thomasplevy 

>>Based on your screenshots it looks like you have a full width view with no sidebar our course outline.

Yes & no. 
Yes - we are using the Launchpad theme & have the screen full width. 
No - we have added a lefthand side bar via BeaverBuilder & in that sidebar have added the Lifter lessons widget.


>>Lessons are 100% inaccessible in LifterLMS when the student doesn't have access. When a student attempts to access a URL (via a direct link) they are redirected to the course where that error message is displayed.

I may not have explained the issue very well. Let me try again:

- Student has completed all the lessons in a section
- Lessons in the next section are not due to be released for a few days
- Student logs out (or session is timed out)
- Student logs in again & goes to Student Dashboard
- From Dashboard, student clicks on course tile
- Student is presented with Course page & clicks on the Lifter button at the bottom of the course page

At this point the error message is displayed.

The Lifter button to navigate from the Course page to the next lesson is attempting to go to a lesson that is not available due to the drip feed rule.

In this scenario, it should be going to the last completed lesson.

Hope this helps.

Tim

**Comment by thomasplevy** [comment]
@wonboyn join our [bug scrub](https://make.lifterlms.com/calendar/events/) tomorrow on [slack](https://lifterlms.com/slack) in `#developers` if you'd like to chat about this in real time too please:

**Comment by wonboyn** [comment]
Hey @thomasplevy ,

>>Now that I'm reading through all this it does feel like this behavior could be adjusted but I'm not positive that adjusting it to move the user to the last available lesson when the next incomplete lesson is not yet available is the correct behavior.

My concern is that the course continue button current behaviour (ie showing the error message & stopping) is preventing users from accessing existing lessons. 

This behaviour has definitely caused confusion for our users.

Whether the button goes to the last completed lesson or a different lesson does not particularly bother me - as long as the button allows the user to access one of the previously completed lessons.

Cheers,

Tim

**Comment by thomasplevy** [comment]
@wonboyn,

This button is not and cannot prevent access to other lessons... I'm sorry if your users are confused but I'm not certain that's the fault of the code powering the button which is why I'm hesitant to just change the existing functionality.

If you do have the lessons listed using the syllabus or outline widget you can see that the behavior of this button has no effect on their ability to access lessons they've already completed.

They can click any already completed lesson from that list and navigate to it that way.

I'm positive this isn't a bug but a question of user experience and the improvement you've recommended feels to me like it will create more support and confusion than it will resolve.

Thanks for your feedback and I'll make some updates to this thread after I discuss with the team tomorrow.

I'd love you to join us for that discussion to advocate your point of view if you're able and willing.

**Comment by wonboyn** [comment]
@thomasplevy 

>>This button is not and cannot prevent access to other lessons..

I must be missing or confusing something really fundamental here...
In the scenario we are facing, the users who have completed all available lessons cick on the course continue button & get the error message **_on the course page_**. They cannot progress from the course page to the lessons.

This is shown in the very first screenshot linked to this issue.

So, from a user experience point of view, **_the button is preventing access_**.


>>If you do have the lessons listed using the syllabus or outline widget

We have added the Lifter "Lesson Sidebar"  on every lesson (via Beaver Builder). Is that not sufficient?

>>I'd love you to join us for that discussion to advocate your point of view if you're able and willing.
Happy to join in depending on time/timezone as I'm based in Australia.
What time is it?

Tim

**Comment by thomasplevy** [comment]
@wonboyn 

> We have added the Lifter "Lesson Sidebar" on every lesson (via Beaver Builder). Is that not sufficient?

The course page -- without any customizations -- serves as a hub for the course. This would list out all of the lessons and with those lists it'll show what they can and cannot access. The contiune button serves as a quick way to jump to the next incomplete lesson.

If you put the list on the lessons that doesn't help this issue where your students have no idea what's available / not available.

With the full default functionality, the UX here isn't as confusing as it is on your site because on your site theres not lesson list so you're correct -- your students cannot get to previous content. This is not the continue button's fault alone. There's a UX issue because your students can't see a list of the available lessons.

The problem is on your course page where they should be able to see the list of lessons (or at least that's the way we've written the defaults with the course syllabus).

> Happy to join in depending on time/timezone as I'm based in Australia.
> What time is it?

Check the event calendar above: https://github.com/gocodebox/lifterlms/issues/728#issuecomment-452035682

**Comment by wonboyn** [comment]
@thomasplevy 

>>The course page -- without any customizations -- serves as a hub for the course. This would list out all of the lessons and with those lists it'll show what they can and cannot access. The contiune button serves as a quick way to jump to the next incomplete lesson.

Ah ok - that was the missing piece of the puzzle for me.
We will add some form of lesson navigation on the course page.

**Comment by thomasplevy** [comment]
After discussing with the team:

Prioritized for potential future enhancement. The best solution we can come up with would be to build a setting (possibly a setting via a filter instead of a database option) which will allow modification of the continue button behavior to redirect to last completed as opposed to next incomplete. We'll monitor support activity as well as this issue to see if there's any other feedback from existing users on the current behavior. Our concern is that adding an option may increase support with users questioning why the option exists and what it's useful for. If we receive more feedback about the current UX we'll adjust the priority accordingly.

An additional enhancement could be to add additional information to the course contiune button (probably via separate block) which would add text defining what's next (Eg: next lesson: lorem ipsum dolor sit (available on 1/1/2019)).

**Comment by thomasplevy** [comment]
Closing due to inactivity. We haven't received any feedback on similar issues since my last comment in 2019. 

Open to reopening in the future for further improvements or feedback but closing for now.
